Sources/PackageCollections/Providers/JSONPackageCollectionProvider.swift

```diff
@@ -118,64 +138,111 @@ struct JSONPackageCollectionProvider: PackageCollectionProvider {
+                            return callback(.failure(
+                                JSONPackageCollectionProviderError
+                                    .invalidResponse(source.url, "Body is empty")
+                            ))
                         }
 
-                        let certPolicyKeys = self.sourceCertPolicy.certificatePolicyKeys(for: source) ?? Self.defaultCertPolicyKeys
-                        self.decodeAndRunSignatureCheck(source: source, data: body, certPolicyKeys: certPolicyKeys, callback: callback)
+                        let certPolicyKeys = self.sourceCertPolicy.certificatePolicyKeys(for: source) ?? Self
+                            .defaultCertPolicyKeys
+                        self.decodeAndRunSignatureCheck(
+                            source: source,
+                            data: body,
+                            certPolicyKeys: certPolicyKeys,
+                            callback: callback
+                        )
                     }
                 }
             }
         }
     }
 
-    private func decodeAndRunSignatureCheck(source: Model.CollectionSource,
-                                            data: Data,
-                                            certPolicyKeys: [CertificatePolicyKey],
-                                            callback: @escaping (Result<Model.Collection, Error>) -> Void) {
+    private func decodeAndRunSignatureCheck(
+        source: Model.CollectionSource,
+        data: Data,
+        certPolicyKeys: [CertificatePolicyKey],
+        callback: @escaping (Result<Model.Collection, Error>) -> Void
+    ) {
         do {
             // This fails if collection is not signed (i.e., no "signature")
             let signedCollection = try self.decoder.decode(JSONModel.SignedCollection.self, from: data)
 
             if source.skipSignatureCheck {
                 // Don't validate signature; set isVerified=false
-                callback(self.makeCollection(from: signedCollection.collection, source: source, signature: Model.SignatureData(from: signedCollection.signature, isVerified: false)))
+                callback(self.makeCollection(
+                    from: signedCollection.collection,
+                    source: source,
+                    signature: Model.SignatureData(from: signedCollection.signature, isVerified: false)
+                ))
             } else if !Self.isSignatureCheckSupported {
                 callback(.failure(StringError("Unsupported platform")))
             } else {
                 // Check the signature
                 let signatureResults = ThreadSafeArrayStore<Result<Void, Error>>()
                 certPolicyKeys.forEach { certPolicyKey in
-                    self.signatureValidator.validate(signedCollection: signedCollection, certPolicyKey: certPolicyKey) { result in
-                        let count = signatureResults.append(result)
+                    Task {
```

Comment:
👀 the key change in this file. The rest is mostly formatting.

Reply:
is the change to run this in a task context, or something more?

Reply:
To run async code in a Task yes. It is inside a loop and each task updates the `signatureResults` array.

Reply:
could we use a task group maybe to help collect the results and make waiting for all the tasks to complete easier? just an idea, not sure if helpful

Reply:
amended https://github.com/apple/swift-package-manager/pull/6609/commits/073d7df9cf4cde7584ba55c4dffaa698e988de41